File: lintcode/python/0149_best_time_to_buy_and_sell_stock.py
#!/usr/bin/python -t

# dp solution, time O(n) space O(1)

class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        # write your code here
        # state: dp[i] max profit when we sell on ith day
        # function: dp[i] = max(dp[i-1]+prices[i]-prices[i-1], prices[i]-prices[i-1])
        # init: dp[0] = 0
        # result: ret = max(ret, dp[i])
        
        n = len(prices)
        if n == 0:
            return 0
        
        max_profit = 0
        min_val = prices[0]
        ret = 0
        
        for i in range(1, n):
            min_val = min(min_val, prices[i-1])
            max_profit = prices[i] - min_val
            ret = max(ret, max_profit)
        return ret


# dp solution, time O(n) space O(n)

class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        # write your code here
        # state: dp[i] max profit when we sell on ith day
        # function: dp[i] = max(dp[i-1]+prices[i]-prices[i-1], prices[i]-prices[i-1])
        # init: dp[0] = 0
        # result: ret = max(ret, dp[i])
        
        n = len(prices)
        if n == 0:
            return 0
        
        dp = [0] * n
        ret = 0
        
        for i in range(1, n):
            dp[i] = max(dp[i-1]+prices[i]-prices[i-1], prices[i]-prices[i-1])
            ret = max(ret, dp[i])
            
        return ret

File: lintcode/golang/0149_best_time_to_buy_and_sell_stock.go
//should correct, but no ac.

/**
 * @param prices: Given an integer array
 * @return: Maximum profit
 */

const MaxUint = ^uint(0) 
const MinUint = 0 
const MaxInt = int(MaxUint >> 1) 
const MinInt = -MaxInt - 1

func maxProfit (prices []int) int {
    // write your code here
    if len(prices) == 0 {
        return 0
    }
    
    ret := 0
    curMin := prices[0]
    
    for index := range prices {
        ret = max(ret, (prices[index] - curMin))
        
        if prices[index] < curMin {
            curMin = prices[index]
        }
    }
    
    return ret
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
